### DISCUSSION TITLE
SETUP COMPONENTS, Cluster or Client?

### DISCUSSION THREAD

**Discussion by cozytk** [mlops-for-all/mlops-for-all.github.io#68] [Q&A] [answered]
안녕하세요! 좋은 가이드 만들어주셔서 잘 따라하고 있었습니다. 진행 도중 kubernetes에 대한 이해가 부족해 막히는 부분이 있어서 질문드립니다.
SETUP COMPONENTS에 해당하는 Kubeflow, MLflow, seldon-core, Prometheus & Grafanas 모두 cluster 환경에 설치하는 것이 맞나요?
Introduction을 보면 kubeflow, MLflow 둘다 클러스터에 설치되는데 SETUP COMPONENTS 파트에 해당하는 helm, kustiomize 등의 명령어는 클라이언트의 software 목록에만 있어서 혼동이 되어 질문드립니다!

**Comment by anencore94** [comment]
안녕하세요! 질문 감사드립니다.
말씀해주신대로 kubeflow, mlflow,... 등의 구성요소는 모두 클러스터에 설치되는 것이 맞습니다.
Helm 과 kustomize 는 위의 구성요소들을 클러스터에 조금 더 쉽게 설치하기 위한 클라이언트 툴이라고 보시면 됩니다. 따라서 이들은 클라이언트 쪽에 설치되는게 맞습니다.

혹시 또 궁금하신 점이 있으시면 문의주시면 감사하겠습니다!

**Comment by Aiden-Jeon** [comment, marked as answer]
안녕하세요!
질문 주신 내용이 맞습니다!

- 클러스터: Kubeflow, MLflow, seldon-core, Prometheus & Grafanas 
- 클라이언트: helm, kustiomize

helm과 kustomize는 쿠버네티스 클러스터에 필요한 kubeflow, mlflow와 같은 컴포넌트를 설치할 때 사용하는 소프트웨어이기 때문에 클러스터를 관리하는 클라이언트에 설치 합니다.

조금 더 내용을 추가해보자면 쿠버네티스의 리소스를 관리하는 방벙은 명시적 방법과 묵시적 방법이 있습니다.
묵시적 방법이란 `kubectl create pod ~~` 과 같이 명령을 통해 리소스를 생성하고 관리하는 방법을 의미합니다.
하지만 이렇게 설치할 경우 쿠버네티스에 있는 것들이 어떻게 생성되었는지 알 수 없어서 유지 보수에 어려움이 있습니다.

이에 반해서 명시적인 방법은 `pod.yaml`과 같이 스펙을 정의한 파일로 관리하는 방법을 의미합니다.
kustomize는 kubeflow를 설치할 때 필요한 컴포넌트를 yaml 형태로 정의한 것을 모아 놓은 레포입니다.

혹시 더 공부하고 싶으시다면 Infrastructure as code (IaC) 키워드를 더 보시면 좋을 것 같습니다!

Helm을 공부하시려면 
docker -> docker-compose -> helm 순서로 알아보시면 됩니다.

**Comment by cozytk** [comment]
@anencore94 @Aiden-Jeon 두 분 모두 친절하게 답변해주시고 더 공부할 키워드도 알려주셔서 진심으로 감사드립니다 :)
